storage: fix batch chunk runtime error

hangvane opened the pull request:
## Relevant Issue (if applicable)
_If there are Issues related to this PullRequest, please list it._

## Details
- Fixed batch chunk runtime error that will cause core dump when running containers built with ``--batch-size`` enabled.
- Improve runtime performance of batch chunk.

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

Desiki-high commented:
/retest

hangvane commented:
/retest

hangvane commented:
/retest

imeoer commented:
cc @jiangliu @ccx1024cc PTAL.

ccx1024cc reviewed (changes requested):
Great Job! Please add e2e test for build/run with batch, which reproduces bugfix in master, but resolved in your pr.

bergwolf commented:
@hangvane Do not leave the commit log empty. Please add a detailed explanation in the commit log that covers:
1. What fails without the patch?
2. How does the patch fix it?

jiangliu commented on storage/src/cache/cachedfile.rs:
```diff
@@ -342,15 +342,28 @@ impl FileCacheEntry {
             for c in extended_chunks.iter() {
                 if !matches!(self.chunk_map.is_ready(c.as_ref()), Ok(true)) {
-                    set.insert(meta.get_zran_index(c.id()));
+                    set.insert(
```
```
let zran_idx = meta.get_zran_index(c.id())
                            .map_err(|e| {
                                error!("Failed to get zran index for chunk {}: {}", c.id(), e)
                            })
                            .unwrap_or(u32::MAX);
       set.insert(zran_index);
```

jiangliu commented on storage/src/cache/cachedfile.rs:
```diff
@@ -342,15 +342,28 @@ impl FileCacheEntry {
+                            .map_err(|e| error!("Failed to get zran index for chunk {}: {}", id, e))
+                            .unwrap_or(u32::MAX),
+                    )
+                {
```
```
if id == first {
    break;
} else {
     match meta.get_zran_index(id) {
             Ok(v) => {
                     if set.contains(v) {
                           break;
                     }
             }
             Err(_e) => break
     }
}
```

jiangliu commented on storage/src/cache/cachedfile.rs:
```diff
@@ -363,7 +376,7 @@ impl FileCacheEntry {
             while end > start {
                 let id = extended_chunks[end].id();
-                if id == last || set.contains(&meta.get_zran_index(id)) {
+                if id == last || set.contains(&meta.get_zran_index(id).unwrap_or(u32::MAX)) {
```
```
if id == last {
    break;
} else {
      match &meta.get_zran_index(id) {
           Ok(v) => ...
           Err(_e) => ....
      }
}

jiangliu commented on storage/src/cache/cachedfile.rs:
```diff
@@ -394,27 +407,44 @@ impl FileCacheEntry {
+            let last_chunk = &chunks[chunks.len() - 1];
+
+            // Get blob meta info iff the chunk is batch chunk.
+            let mut meta: Option<Arc<BlobCompressionContextInfo>> = None;
```
How about ?
```
match (first_chunk.is_batch(), last_chunk.is_batch()) {
   (true, true) => ...
   (true, false) => ...
   (false, true) =>...
   (false, false) => ...
}
```

jiangliu replied:
A helper function as below should simplify the implementation.

hangvane replied:
Helper `BlobCCI` is added for simplification.

jiangliu commented on storage/src/cache/cachedfile.rs:
```diff
@@ -1095,10 +1133,38 @@ impl FileCacheEntry {
+                let next_chunk = &region.chunks[idx + 1];
+
+                // Lazy load blob meta info if needed.
+                if (pre_chunk.is_batch() || next_chunk.is_batch()) && meta.is_none() {
```
There are repeated code patterns here which makes implementation a little complex. It would be great to use a helper function here such as:
```
struct BlobCCI {
         meta: Option<Arc<BlobCompressionContextInfo>>,
}

impl BlobCCI {
      fn get_chunk_compressed_offset(chunk: &ChunkInfo) -> Result<u64) {
            xxxxx
      }
}
```

hangvane replied:
The helper `BlobCCI` is implemented.

jiangliu commented on builder/src/core/node.rs:
```diff
@@ -388,15 +394,18 @@ impl Node {
         chunk_data: &[u8],
         chunk: &mut ChunkWrapper,
-    ) -> Result<Option<BlobChunkInfoV2Ondisk>> {
+    ) -> Result<(Option<BlobChunkInfoV2Ondisk>, Option<u32>)> {
```
How about returning `Result<(Option<BlobChunkInfoV2Ondisk>, u32)>`, which may help to simplify the implementation?

hangvane replied:
fixed.

jiangliu commented on storage/src/meta/chunk_info_v2.rs:
```diff
@@ -201,45 +212,95 @@ impl BlobMetaChunkInfo for BlobChunkInfoV2Ondisk {
             let ctx = &state.zran_info_array[index];
-            if self.get_zran_offset() >= ctx.out_size()
-                || self.get_zran_offset() + self.uncompressed_size() > ctx.out_size()
+            if self.get_zran_offset()? >= ctx.out_size()
```
```
let zran_offset = self.get_zran_offset()?;
```

hangvane replied:
Fixed.

jiangliu commented on storage/src/meta/chunk_info_v2.rs:
```diff
@@ -201,45 +212,95 @@ impl BlobMetaChunkInfo for BlobChunkInfoV2Ondisk {
+            }
+        }
+
+        if state.blob_features & BlobFeatures::BATCH.bits() == 0 && self.is_batch() {
```
```
if self.is_batch() {
    if state.blob_features & BlobFeatures::BATCH.bits() == 0 {
     ....
    } else {
      .....
     }
}
```

hangvane replied:
Changed.

imeoer commented:
@hangvane What's the status of this PR, can we push it forward?

hangvane commented on builder/src/core/node.rs:
```diff
@@ -310,17 +310,23 @@ impl Node {
             chunk.set_blob_index(blob_index);
             chunk.set_index(chunk_index);
             chunk.set_file_offset(file_offset);
+            let mut dumped_size = chunk.compressed_size();
```
Seems better to leave `dumped_size` as it is. Because `dumped_size` have an initial value, `chunk.compressed_size()`, which is non-zero for zran-like referenced type. So `dumped_size` is non-empty and can't be wrapped by `Option`.

imeoer commented:
@jiangliu Please take a look this again, thanks!

hangvane committed:
rafs: move `compressed_offset` from `BatchInflateContext` to chunk info for batch chunks.

1. `compressed_offset` is used for build-time and runtime sorting for chunk info.
So we move `compressed_offset` from `BatchInflateContext` to chunk info for batch chunks.

2. the `compressed_size` for the blobs in batch mode is not correctly set.
We thus fix it by setting the value of `dumped_size`.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: fix the Error type to log the error correctly

Currently, many error are output as `os error 22` lossing customized log info.
So we change the Error type for correctly output and log the error info
as what we expected.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: fix the read amplification for batch chunks.

Read amplification for batch chunk is not correctly implemented that may crash.
The read amplification is rewrited to fix this bug.
A unit test for read amplification is also added for covering this code.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
rafs: add `is_batch()` for `BlobChunkInfo`

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: add validation and unit test for batch chunks

1. Add the validation for batch chunks.
2. Add unit test for `BatchInflateContext`.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: introduce `BlobCCI` for reading batch chunk info

`BlobCompressionContextInfo` is need to read batch chunk info.
`BlobCCI` is introduced for simplifying the code,
and decrease the times of getting this context by lazy loading.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
action: add convert test for batch chunk

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: fix the tiny prefetch request for batch chunks

By passing the chunk continuous check, and correctly sort batch chunks,
the prefetch request will not be interrupted by batch chunks anymore.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane commented:
> Great Job! Please add e2e test for build/run with batch, which reproduces bugfix in master, but resolved in your pr.

E2E tests have been added to master,
https://github.com/dragonflyoss/nydus/blob/a4ec9b806167786eb064da20c55e0ec34653b71d/smoke/tests/native_layer_test.go#L43-L45

and may randomly fail E2E tests. But the logs are expired and cannot be traced.
https://github.com/dragonflyoss/nydus/actions/runs/5452176504

Desiki-high commented:
@ccx1024cc Please take a look at this again, thanks!
